pulsar-broker/src/main/java/org/apache/pulsar/broker/service/InMemoryAndPreventCycleFilterRedeliveryTracker.java

```diff
@@ -0,0 +1,173 @@
+        if (redeliveryStartAt == null || comparePosition(redeliveryStartAt, position) == 0
+                || comparePosition(redeliveryStartAt, markDeletedPosition) >= 0) {
+            cleanEarliestInformation();
+        }
+    }
+
+    @Override
+    public void clear() {
+        super.clear();
+        cleanEarliestInformation();
+    }
+
+    public boolean isConsumerPaused(Consumer consumer) {
+        if (consumer == null) {
+            return false;
+        }
+        PauseConsumerInformation pauseConsumerInformation = pausedConsumers.get(consumer);
+        if (pauseConsumerInformation == null) {
+            return false;
+        }
+        if (!pauseConsumerInformation.isValid(redeliveryStartAt)) {
+            pausedConsumers.remove(consumer);
+            return false;
+        }
+        return true;
+    }
+
+    public int pausedConsumerCount() {
+        return (int) pausedConsumers.keySet().stream().map(this::isConsumerPaused).count();
+    }
+
+    private void cleanEarliestInformation() {
+        redeliveryStartAt = null;
+        // If consumer has been closed, remove this consumer.
+        List<Consumer> closedConsumers = earliestEntryRedeliveryCountMapping.keySet().stream()
+                .filter(Consumer::isClosed).toList();
+        closedConsumers.forEach(earliestEntryRedeliveryCountMapping::remove);
+        // Just reset counter to 0, because value will be removed if consumer is closed.
+        earliestEntryRedeliveryCountMapping.values().forEach(i -> i.set(0));
+        pausedConsumers.clear();
+    }
+
+    private static int comparePosition(Position pos1, Position pos2) {
+        int ledgerCompare = Long.compare(pos1.getLedgerId(), pos2.getLedgerId());
+        if (ledgerCompare != 0) {
+            return ledgerCompare;
+        }
+        return Long.compare(pos1.getEntryId(), pos2.getEntryId());
+    }
+
+    private static class PauseConsumerInformation {
+
+        private final Position cantConsumedPosition;
+
+        private final long pauseTime;
+
+        private PauseConsumerInformation(Position cantConsumedPosition) {
+            this.cantConsumedPosition = cantConsumedPosition;
+            this.pauseTime = System.currentTimeMillis();
+        }
+
+        /**
+         * If "do pause consumer" and {@link #cleanEarliestInformation} concurrently, it is possible to pause
+         * consumer that no longer needs to be paused. The way to distinguish is to determine whether
+         * "cantConsumedPosition equal to currentRedeliveryStartAt".
+         * {@link PauseConsumerInformation#isValid(Position)} resolved this problem.
+         */
+        boolean isValid(Position currentRedeliveryStartAt) {
+            if (cantConsumedPosition != currentRedeliveryStartAt) {
+                return false;
+            }
+            // Automatically becomes invalid after 1s, because users may use time to filter the Entry.
+            return System.currentTimeMillis() - pauseTime < 1000;
```

Comment:
this should be configurable, maybe next to ServiceConfiguration#dispatcherEntryFilterRescheduledMessageDelay ?

Reply:
Already fixed. Could you please help me check whether DOC is suitable?